File: src/contests/cg/puzzle/php/telephone-numbers.php
<?php
// 🎮 CodinGame Puzzle - telephone-numbers
// https://www.codingame.com/training/medium/telephone-numbers

$n = (int) trim(fgets(STDIN));
$root = [];
$cableUnits = 0;
for ($i = 0; $i < $n; $i++) {
    $number = trim(fgets(STDIN));
    $node = &$root;
    $len = strlen($number);
    for ($j = 0; $j < $len; $j++) {
        $d = $number[$j];
        if (!isset($node[$d])) {
            $node[$d] = [];
            $cableUnits++;
        }
        $node = &$node[$d];
    }
    unset($node);
}
echo $cableUnits . "\n";

File: src/contests/cg/puzzle/lua/telephone-numbers.lua
-- 🎮 CodinGame Puzzle - telephone-numbers
-- https://www.codingame.com/training/medium/telephone-numbers

local n = tonumber(io.read())
local root = {}
local cable_units = 0
for _ = 1, n do
  local number = (io.read():gsub("%s+$", ""))
  local node = root
  for d in number:gmatch(".") do
    if node[d] == nil then
      node[d] = {}
      cable_units = cable_units + 1
    end
    node = node[d]
  end
end
print(cable_units)
